lib/test/cdo/test_active_job_backend.rb

```diff
@@ -0,0 +1,333 @@
+      Cdo::ActiveJobBackend.expects(:start_n_workers).returns(5).in_sequence(sequence)
+
+      n_workers_running = Cdo::ActiveJobBackend.restart_workers_internal(5, n_batches: 1)
+      assert_equal 5, n_workers_running
+    end
+
+    it 'can start 5 workers in 2 batches' do
+      sequence = Mocha::Sequence.new('stop then start')
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89890, 89892, 89894]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(3, initial_worker_index: 0).returns(3).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89936, 89938]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(2, initial_worker_index: 3).returns(2).in_sequence(sequence)
+
+      n_workers_running = Cdo::ActiveJobBackend.restart_workers_internal(5, n_batches: 2)
+      assert_equal 5, n_workers_running
+    end
+
+    it 'can start 5 workers when only 4 are running' do
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_4_stale_workers)
+
+      sequence = Mocha::Sequence.new('stop then start')
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89890, 89892,]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(2, initial_worker_index: 0).returns(2).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89894, 89936]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(2, initial_worker_index: 2).returns(2).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(1, initial_worker_index: 4).returns(1).in_sequence(sequence)
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_5_fresh_workers).in_sequence(sequence)
+
+      n_workers_running = Cdo::ActiveJobBackend.restart_workers_internal(5, n_batches: 2)
+      assert_equal 5, n_workers_running
+    end
+
+    it 'can start 4 workers when 5 are running' do
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_5_stale_workers)
+
+      sequence = Mocha::Sequence.new('stop then start')
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89890, 89892, 89894]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(3, initial_worker_index: 0).returns(3).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:stop_workers).with {|pids, *| pids == [89936, 89938]}.in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:start_n_workers).with(1, initial_worker_index: 3).returns(2).in_sequence(sequence)
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_4_fresh_workers).in_sequence(sequence)
+
+      n_workers_running = Cdo::ActiveJobBackend.restart_workers_internal(4, n_batches: 2)
+      assert_equal 4, n_workers_running
+    end
```

Comment:
> Do we need a test for the scenario where n_workers_to_start is less than the batch size?

That's the test right above this one ;-)

- Just to be clear: n_batches is fixed by us (5 on prod) and determines capacity loss during deploy (inverse of n_batches).
- n_workers_restarted_in_each_batch is determined based on number of current processes running (not num being launched), as you infer. Its based on num workers running because what matters for maintaining capacity (the point of batches) is stopping of existing workers not starting of new ones. That is, if it was just about starting new workers, we'd start them all in one giant batch. Batch's reason to exist is 100% stopping (and then replacing) workers.